File: src/Application Specific/Classes/Entity Collections/Jobs/JobOfWork/JobOfWorkSQL.vb
Imports System.Collections.Generic
Imports System.Data.SqlClient
Imports System.Linq
Imports FSM.Entity.Sys

Namespace Entity

    Namespace JobOfWorks

        Public Class JobOfWorkSQL
            Private _database As Sys.Database

            Public Sub New(ByVal database As Sys.Database)
                _database = database
            End Sub

#Region "Functions"

            Public Sub Delete(ByVal JobOfWorkID As Integer)
                _database.ClearParameter()
                _database.AddParameter("@JobOfWorkID", JobOfWorkID, True)

                _database.ExecuteSP_NO_Return("JobOfWork_Delete")
            End Sub

            Public Sub Delete(ByVal JobOfWorkID As Integer, ByVal trans As SqlClient.SqlTransaction)

                Dim Command As SqlClient.SqlCommand = New SqlClient.SqlCommand

                Command.CommandText = "JobOfWork_Delete"
                Command.CommandType = CommandType.StoredProcedure
                Command.Transaction = trans
                Command.Connection = trans.Connection

                Command.Parameters.Add(New System.Data.SqlClient.SqlParameter("@JobOfWorkID", JobOfWorkID))

                Command.ExecuteNonQuery()

            End Sub

            Public Function [Insert](ByVal oJobOfWork As JobOfWork, ByVal trans As SqlClient.SqlTransaction) As JobOfWork

                Dim Command As SqlClient.SqlCommand = New SqlClient.SqlCommand

                Command.CommandText = "JobOfWork_Insert"
                Command.CommandType = CommandType.StoredProcedure
                Command.Transaction = trans
                Command.Connection = trans.Connection

                Command.Parameters.Add(New System.Data.SqlClient.SqlParameter("@JobID", oJobOfWork.JobID))
                Command.Parameters.Add(New System.Data.SqlClient.SqlParameter("@Deleted", oJobOfWork.Deleted))
                Command.Parameters.Add(New System.Data.SqlClient.SqlParameter("@PONumber", oJobOfWork.PONumber))
                Command.Parameters.Add(New System.Data.SqlClient.SqlParameter("@Status", oJobOfWork.Status))
                Command.Parameters.Add(New System.Data.SqlClient.SqlParameter("@Priority", oJobOfWork.Priority))
                Command.Parameters.Add(New System.Data.SqlClient.SqlParameter("@QualificationID", oJobOfWork.QualificationID))
                If oJobOfWork.PriorityDateSet = Date.MinValue Then
                    Command.Parameters.Add(New System.Data.SqlClient.SqlParameter("@PriorityDateSet", DBNull.Value))
                Else
                    Command.Parameters.Add(New System.Data.SqlClient.SqlParameter("@PriorityDateSet", oJobOfWork.PriorityDateSet))
                End If

                oJobOfWork.SetJobOfWorkID = Entity.Sys.Helper.MakeIntegerValid(Command.ExecuteScalar)
                oJobOfWork.Exists = True

                For Each jobItem As Entity.JobItems.JobItem In oJobOfWork.JobItems
                    jobItem.SetJobOfWorkID = oJobOfWork.JobOfWorkID
                    jobItem = _database.JobItems.Insert(jobItem, trans)
                Next

                For Each engineerVisit As Entity.EngineerVisits.EngineerVisit In oJobOfWork.EngineerVisits
                    engineerVisit.SetJobOfWorkID = oJobOfWork.JobOfWorkID
                    engineerVisit = _database.EngineerVisits.Insert(engineerVisit, oJobOfWork.JobID, trans)
                Next

                Return oJobOfWork
            End Function

            Public Function [Insert](ByVal oJobOfWork As JobOfWork) As JobOfWork

                _database.ClearParameter()
                _database.AddParameter("@JobID", oJobOfWork.JobID, True)
                _database.AddParameter("@Deleted", oJobOfWork.Deleted, True)
                _database.AddParameter("@PONumber ", oJobOfWork.PONumber, True)
                _database.AddParameter("@Status", oJobOfWork.Status, True)
                _database.AddParameter("@Priority ", oJobOfWork.Priority, True)
                _database.AddParameter("@QualificationID ", oJobOfWork.QualificationID, True)

                If oJobOfWork.PriorityDateSet = Date.MinValue Or oJobOfWork.PriorityDateSet = Nothing Then
                    _database.AddParameter("@PriorityDateSet ", DBNull.Value, True)
                Else
                    _database.AddParameter("@PriorityDateSet ", oJobOfWork.PriorityDateSet, True)
                End If

                oJobOfWork.SetJobOfWorkID = Entity.Sys.Helper.MakeIntegerValid(_database.ExecuteSP_OBJECT("JobOfWork_Insert"))
                oJobOfWork.Exists = True

                For Each jobItem As Entity.JobItems.JobItem In oJobOfWork.JobItems
                    jobItem.SetJobOfWorkID = oJobOfWork.JobOfWorkID
                    jobItem = _database.JobItems.Insert(jobItem)
                Next

                For Each engineerVisit As Entity.EngineerVisits.EngineerVisit In oJobOfWork.EngineerVisits
                    engineerVisit.SetJobOfWorkID = oJobOfWork.JobOfWorkID
                    engineerVisit = _database.EngineerVisits.Insert(engineerVisit, oJobOfWork.JobID)
                Next

                Return oJobOfWork
            End Function

            Public Sub Update_PONumber(ByVal oJobOfWork As JobOfWork)
                If oJobOfWork.PriorityDateSet = Date.MinValue Then
                    oJobOfWork.PriorityDateSet = DateTime.Now
                End If
                _database.ClearParameter()
                _database.AddParameter("@JobOfWorkID", oJobOfWork.JobOfWorkID, True)
                _database.AddParameter("@PONumber ", oJobOfWork.PONumber, True)
                _database.AddParameter("@Priority", oJobOfWork.Priority, True)
                _database.AddParameter("@PriorityDateSet", oJobOfWork.PriorityDateSet, True)
                _database.AddParameter("@QualificationID", oJobOfWork.QualificationID, True)

                _database.ExecuteSP_NO_Return("JobOfWork_Update_PONumber")
            End Sub

            Public Sub Update_PONumber(ByVal oJobOfWork As JobOfWork, ByVal trans As SqlClient.SqlTransaction)

                Dim Command As SqlClient.SqlCommand = New SqlClient.SqlCommand

                Command.CommandText = "JobOfWork_Update_PONumber"
                Command.CommandType = CommandType.StoredProcedure
                Command.Transaction = trans
                Command.Connection = trans.Connection

                Command.Parameters.Add(New System.Data.SqlClient.SqlParameter("@JobOfWorkID", oJobOfWork.JobOfWorkID))
                Command.Parameters.Add(New System.Data.SqlClient.SqlParameter("@PONumber", oJobOfWork.PONumber))
                Command.Parameters.Add(New System.Data.SqlClient.SqlParameter("@QualificationID", oJobOfWork.QualificationID))

                Command.Parameters.Add(New System.Data.SqlClient.SqlParameter("@Priority", oJobOfWork.Priority))
                If oJobOfWork.PriorityDateSet = Date.MinValue Then
                    Command.Parameters.Add(New System.Data.SqlClient.SqlParameter("@PriorityDateSet", DBNull.Value))
                Else
                    Command.Parameters.Add(New System.Data.SqlClient.SqlParameter("@PriorityDateSet", oJobOfWork.PriorityDateSet))
                End If

                Command.ExecuteNonQuery()

            End Sub

            Public Sub Update_Status(ByVal JobOfWorkID As Integer, ByVal Status As Integer)

                _database.ClearParameter()
                _database.AddParameter("@JobOfWorkID", JobOfWorkID, True)
                _database.AddParameter("@Status ", Status, True)

                _database.ExecuteSP_NO_Return("JobOfWork_Update_Status")

            End Sub

            Public Sub Update_Priority(ByVal Jow As Entity.JobOfWorks.JobOfWork)
                _database.ClearParameter()
                _database.AddParameter("@JobOfWorkID", Jow.JobOfWorkID, True)
                _database.AddParameter("@Priority ", Jow.Priority, True)
                _database.AddParameter("@PriorityDateSet ", Jow.PriorityDateSet, True)
                _database.AddParameter("@QualificationID ", Jow.QualificationID, True)

                _database.ExecuteSP_NO_Return("JobOfWork_Update_Priority")

            End Sub

            Public Function JobOfWork_Get_For_Job(ByVal JobID As Integer) As DataView
                _database.ClearParameter()
                _database.AddParameter("@JobID", JobID, True)

                Dim dt As DataTable = _database.ExecuteSP_DataTable("JobOfWork_Get_For_Job")
                dt.TableName = Entity.Sys.Enums.TableNames.tblJobOfWork.ToString
                Return New DataView(dt)
            End Function

            Public Function JobOfWork_Get_For_Job(ByVal JobID As Integer, ByVal trans As SqlClient.SqlTransaction) As DataView

                Dim Command As SqlClient.SqlCommand = New SqlClient.SqlCommand

                Command.CommandText = "JobOfWork_Get_For_Job"
                Command.CommandType = CommandType.StoredProcedure
                Command.Transaction = trans
                Command.Connection = trans.Connection

                Command.Parameters.AddWithValue("@JobID", JobID)

                Dim Adapter As New SqlClient.SqlDataAdapter(Command)
                Dim returnDS As New DataSet
                Adapter.Fill(returnDS)

                Dim dt As DataTable = returnDS.Tables(0)

                dt.TableName = Entity.Sys.Enums.TableNames.tblJobOfWork.ToString
                Return New DataView(dt)

            End Function

            Public Function JobOfWork_Get_ForEngineerVisitID(ByVal engineerVisitId As Integer) As DataTable
                Dim dt As New DataTable
                Dim command As New SqlCommand("JobOfWork_Get_ForEngineerVisitID", New SqlConnection(_database.ServerConnectionString))
                Try
                    command.CommandType = CommandType.StoredProcedure
                    command.Parameters.AddWithValue("@EngineerVisitID", engineerVisitId)

                    dt = _database.ExecuteCommand_DataTable(command)
                    dt.TableName = "tblJobOfWork"

                    Return dt
                Catch ex As Exception
                    Return dt
                Finally
                    command.Connection.Close()
                End Try
            End Function

            Public Function JobOfWork_Get(ByVal JobOfWorkID As Integer) As JobOfWorks.JobOfWork

                _database.ClearParameter()
                _database.AddParameter("@JobOfWorkID", JobOfWorkID, True)

                Dim dt As DataTable = _database.ExecuteSP_DataTable("JobOfWork_Get")
                Dim jobOfWork As JobOfWorks.JobOfWork = Nothing

                If dt.Rows.Count > 0 Then

                    jobOfWork = New JobOfWork

                    With dt.Rows(0)
                        jobOfWork.IgnoreExceptionsOnSetMethods = True
                        jobOfWork.Exists = True
                        jobOfWork.SetJobOfWorkID = .Item("JobOFWorkID")
                        jobOfWork.SetJobID = .Item("JobID")
                        jobOfWork.SetPONumber = .Item("PONumber")
                        jobOfWork.SetDeleted = .Item("Deleted")
                        jobOfWork.SetStatus = Sys.Helper.MakeIntegerValid(.Item("Status"))
                        jobOfWork.SetPriority = Sys.Helper.MakeIntegerValid(.Item("Priority"))
                        jobOfWork.PriorityDateSet = Sys.Helper.MakeDateTimeValid(.Item("PriorityDateSet"))
                        If dt.Rows(0).Table.Columns.Contains("QualificationID") Then jobOfWork.SetQualificationID = Sys.Helper.MakeIntegerValid(.Item("QualificationID"))
                    End With

                End If

                Return jobOfWork

            End Function

            Public Function JobOfWork_Get_As_Object(ByVal jobOfWorkID As Integer) As JobOfWork
                _database.ClearParameter()
                _database.AddParameter("@JobOfWorkID", jobOfWorkID, True)

                Dim dt As DataTable = _database.ExecuteSP_DataTable("JobOfWork_Get")
                Dim jobOfWork As JobOfWork = Nothing

                If dt.Rows.Count > 0 Then

                    jobOfWork = New JobOfWork

                    With dt.Rows(0)
                        jobOfWork.IgnoreExceptionsOnSetMethods = True
                        jobOfWork.Exists = True
                        jobOfWork.SetJobOfWorkID = .Item("JobOFWorkID")
                        jobOfWork.SetJobID = .Item("JobID")
                        jobOfWork.SetPONumber = .Item("PONumber")
                        jobOfWork.SetDeleted = .Item("Deleted")
                        jobOfWork.SetStatus = Sys.Helper.MakeIntegerValid(.Item("Status"))
                        jobOfWork.SetPriority = Sys.Helper.MakeIntegerValid(.Item("Priority"))
                        jobOfWork.PriorityDateSet = Sys.Helper.MakeDateTimeValid(.Item("PriorityDateSet"))
                        If dt.Rows(0).Table.Columns.Contains("QualificationID") Then jobOfWork.SetQualificationID = Sys.Helper.MakeIntegerValid(.Item("QualificationID"))
                        jobOfWork.JobItems = _database.JobItems.JobOfWork_Get_For_Job_Of_Work_As_Objects(jobOfWork.JobOfWorkID)
                        jobOfWork.EngineerVisits = _database.EngineerVisits.EngineerVisits_Get_For_Job_Of_Work_As_Objects_Light(jobOfWork.JobOfWorkID)
                    End With

                End If
                Return jobOfWork
            End Function

            Public Function JobOfWork_Get_For_Job_As_Objects(ByVal JobID As Integer, ByVal trans As SqlClient.SqlTransaction) As ArrayList

                Dim jobOfWorks As New ArrayList
                For Each row As DataRow In JobOfWork_Get_For_Job(JobID, trans).Table.Rows
                    Dim jobOfWork As New JobOfWorks.JobOfWork
                    jobOfWork.IgnoreExceptionsOnSetMethods = True
                    jobOfWork.Exists = True
                    jobOfWork.SetJobOfWorkID = row.Item("JobOFWorkID")
                    jobOfWork.SetJobID = row.Item("JobID")
                    jobOfWork.SetPONumber = row.Item("PONumber")
                    jobOfWork.SetDeleted = row.Item("Deleted")
                    jobOfWork.SetStatus = Sys.Helper.MakeIntegerValid(row.Item("Status"))
                    jobOfWork.SetPriority = Sys.Helper.MakeIntegerValid(row.Item("Priority"))
                    jobOfWork.PriorityDateSet = Sys.Helper.MakeDateTimeValid(row.Item("PriorityDateSet"))
                    If row.Table.Columns.Contains("QualificationID") Then jobOfWork.SetQualificationID = Sys.Helper.MakeIntegerValid(row.Item("QualificationID"))
                    jobOfWork.JobItems = _database.JobItems.JobOfWork_Get_For_Job_Of_Work_As_Objects(jobOfWork.JobOfWorkID, trans)
                    jobOfWork.EngineerVisits = _database.EngineerVisits.EngineerVisits_Get_For_Job_Of_Work_As_Objects(jobOfWork.JobOfWorkID, trans)
                    jobOfWorks.Add(jobOfWork)
                Next

                Return jobOfWorks
            End Function

            Public Function JobOfWork_Get_For_Job_As_Objects(ByVal JobID As Integer) As ArrayList
                Dim jobOfWorks As New ArrayList
                For Each row As DataRow In JobOfWork_Get_For_Job(JobID).Table.Rows
                    Dim jobOfWork As New JobOfWorks.JobOfWork
                    jobOfWork.IgnoreExceptionsOnSetMethods = True
                    jobOfWork.Exists = True
                    jobOfWork.SetJobOfWorkID = row.Item("JobOFWorkID")
                    jobOfWork.SetJobID = row.Item("JobID")
                    jobOfWork.SetPONumber = row.Item("PONumber")
                    jobOfWork.SetDeleted = row.Item("Deleted")
                    jobOfWork.SetStatus = Sys.Helper.MakeIntegerValid(row.Item("Status"))
                    jobOfWork.SetPriority = Sys.Helper.MakeIntegerValid(row.Item("Priority"))
                    jobOfWork.PriorityDateSet = Sys.Helper.MakeDateTimeValid(row.Item("PriorityDateSet"))
                    If row.Table.Columns.Contains("QualificationID") Then jobOfWork.SetQualificationID = Sys.Helper.MakeIntegerValid(row.Item("QualificationID"))
                    jobOfWork.JobItems = _database.JobItems.JobOfWork_Get_For_Job_Of_Work_As_Objects(jobOfWork.JobOfWorkID)
                    jobOfWork.EngineerVisits = _database.EngineerVisits.EngineerVisits_Get_For_Job_Of_Work_As_Objects_Light(jobOfWork.JobOfWorkID)

                    jobOfWorks.Add(jobOfWork)
                Next

                Return jobOfWorks
            End Function

            Public Sub Update_PONumberByJobOfWorkID(ByVal jobOfWorkId As Integer, ByVal poNumber As String)
                _database.ClearParameter()
                _database.AddParameter("@JobOfWorkID ", jobOfWorkId, True)
                _database.AddParameter("@PONumber ", poNumber, True)
                _database.ExecuteSP_NO_Return("JobOfWork_Update_PONumberByJobOfWorkID")
            End Sub

            Public Function [Get_ByJobId](ByVal jobId As Integer) As List(Of JobOfWork)
                _database.ClearParameter()
                _database.AddParameter("@JobID", jobId, True)
                Dim dt As DataTable = _database.ExecuteSP_DataTable("JobOfWork_Get_For_Job")
                If dt IsNot Nothing AndAlso dt.Rows.Count > 0 Then

                    Dim jobOfWorks As List(Of JobOfWork) =
                        (From x In dt.AsEnumerable() Select New JobOfWork With {
                            .Exists = True,
                            .SetJobOfWorkID = Helper.MakeIntegerValid(x("JobOfWorkId")),
                            .SetJobID = Helper.MakeIntegerValid(x("JobId")),
                            .SetPONumber = Helper.MakeStringValid(x("PONumber")),
                            .SetPriority = Helper.MakeIntegerValid(x("Priority")),
                            .SetStatus = Helper.MakeIntegerValid(x("Status")),
                            .PriorityDateSet = Helper.MakeDateTimeValid(x("PriorityDateSet")),
                            .SetQualificationID = Helper.MakeIntegerValid(x("QualificationID"))
                        }).ToList()
                    Return jobOfWorks
                Else : Return Nothing
                End If
            End Function

#End Region

        End Class

    End Namespace

End Namespace

File: src/Application Specific/Classes/Entity Collections/Jobs/JobOfWork/JobOfWorkSQL.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Data.SqlClient;
using System.Linq;
using FSM.Entity.Sys;
using Microsoft.VisualBasic.CompilerServices;

namespace FSM.Entity
{
    namespace JobOfWorks
    {
        public class JobOfWorkSQL
        {
            private Database _database;

            public JobOfWorkSQL(Database database)
            {
                _database = database;
            }

            /* TODO ERROR: Skipped RegionDirectiveTrivia */

            public void Delete(int JobOfWorkID)
            {
                _database.ClearParameter();
                _database.AddParameter("@JobOfWorkID", JobOfWorkID, true);
                _database.ExecuteSP_NO_Return("JobOfWork_Delete");
            }

            public void Delete(int JobOfWorkID, SqlTransaction trans)
            {
                var Command = new SqlCommand();
                Command.CommandText = "JobOfWork_Delete";
                Command.CommandType = CommandType.StoredProcedure;
                Command.Transaction = trans;
                Command.Connection = trans.Connection;
                Command.Parameters.Add(new SqlParameter("@JobOfWorkID", JobOfWorkID));
                Command.ExecuteNonQuery();
            }

            public JobOfWork Insert(JobOfWork oJobOfWork, SqlTransaction trans)
            {
                var Command = new SqlCommand();
                Command.CommandText = "JobOfWork_Insert";
                Command.CommandType = CommandType.StoredProcedure;
                Command.Transaction = trans;
                Command.Connection = trans.Connection;
                Command.Parameters.Add(new SqlParameter("@JobID", oJobOfWork.JobID));
                Command.Parameters.Add(new SqlParameter("@Deleted", oJobOfWork.Deleted));
                Command.Parameters.Add(new SqlParameter("@PONumber", oJobOfWork.PONumber));
                Command.Parameters.Add(new SqlParameter("@Status", oJobOfWork.Status));
                Command.Parameters.Add(new SqlParameter("@Priority", oJobOfWork.Priority));
                Command.Parameters.Add(new SqlParameter("@QualificationID", oJobOfWork.QualificationID));
                if (oJobOfWork.PriorityDateSet == DateTime.MinValue)
                {
                    Command.Parameters.Add(new SqlParameter("@PriorityDateSet", DBNull.Value));
                }
                else
                {
                    Command.Parameters.Add(new SqlParameter("@PriorityDateSet", oJobOfWork.PriorityDateSet));
                }

                oJobOfWork.SetJobOfWorkID = Helper.MakeIntegerValid(Command.ExecuteScalar());
                oJobOfWork.Exists = true;
                foreach (JobItems.JobItem jobItem in oJobOfWork.JobItems)
                {
                    jobItem.SetJobOfWorkID = oJobOfWork.JobOfWorkID;
                    _database.JobItems.Insert(jobItem, trans);
                }

                foreach (EngineerVisits.EngineerVisit engineerVisit in oJobOfWork.EngineerVisits)
                {
                    engineerVisit.SetJobOfWorkID = oJobOfWork.JobOfWorkID;
                    _database.EngineerVisits.Insert(engineerVisit, oJobOfWork.JobID, trans);
                }

                return oJobOfWork;
            }

            public JobOfWork Insert(JobOfWork oJobOfWork)
            {
                _database.ClearParameter();
                _database.AddParameter("@JobID", oJobOfWork.JobID, true);
                _database.AddParameter("@Deleted", oJobOfWork.Deleted, true);
                _database.AddParameter("@PONumber ", oJobOfWork.PONumber, true);
                _database.AddParameter("@Status", oJobOfWork.Status, true);
                _database.AddParameter("@Priority ", oJobOfWork.Priority, true);
                _database.AddParameter("@QualificationID ", oJobOfWork.QualificationID, true);
                if (oJobOfWork.PriorityDateSet == DateTime.MinValue | oJobOfWork.PriorityDateSet == default)
                {
                    _database.AddParameter("@PriorityDateSet ", DBNull.Value, true);
                }
                else
                {
                    _database.AddParameter("@PriorityDateSet ", oJobOfWork.PriorityDateSet, true);
                }

                oJobOfWork.SetJobOfWorkID = Helper.MakeIntegerValid(_database.ExecuteSP_OBJECT("JobOfWork_Insert"));
                oJobOfWork.Exists = true;
                foreach (JobItems.JobItem jobItem in oJobOfWork.JobItems)
                {
                    jobItem.SetJobOfWorkID = oJobOfWork.JobOfWorkID;
                    _database.JobItems.Insert(jobItem);
                }

                foreach (EngineerVisits.EngineerVisit engineerVisit in oJobOfWork.EngineerVisits)
                {
                    engineerVisit.SetJobOfWorkID = oJobOfWork.JobOfWorkID;
                    _database.EngineerVisits.Insert(engineerVisit, oJobOfWork.JobID);
                }

                return oJobOfWork;
            }

            public void Update_PONumber(JobOfWork oJobOfWork)
            {
                if (oJobOfWork.PriorityDateSet == DateTime.MinValue)
                {
                    oJobOfWork.PriorityDateSet = DateTime.Now;
                }

                _database.ClearParameter();
                _database.AddParameter("@JobOfWorkID", oJobOfWork.JobOfWorkID, true);
                _database.AddParameter("@PONumber ", oJobOfWork.PONumber, true);
                _database.AddParameter("@Priority", oJobOfWork.Priority, true);
                _database.AddParameter("@PriorityDateSet", oJobOfWork.PriorityDateSet, true);
                _database.AddParameter("@QualificationID", oJobOfWork.QualificationID, true);
                _database.ExecuteSP_NO_Return("JobOfWork_Update_PONumber");
            }

            public void Update_PONumber(JobOfWork oJobOfWork, SqlTransaction trans)
            {
                var Command = new SqlCommand();
                Command.CommandText = "JobOfWork_Update_PONumber";
                Command.CommandType = CommandType.StoredProcedure;
                Command.Transaction = trans;
                Command.Connection = trans.Connection;
                Command.Parameters.Add(new SqlParameter("@JobOfWorkID", oJobOfWork.JobOfWorkID));
                Command.Parameters.Add(new SqlParameter("@PONumber", oJobOfWork.PONumber));
                Command.Parameters.Add(new SqlParameter("@QualificationID", oJobOfWork.QualificationID));
                Command.Parameters.Add(new SqlParameter("@Priority", oJobOfWork.Priority));
                if (oJobOfWork.PriorityDateSet == DateTime.MinValue)
                {
                    Command.Parameters.Add(new SqlParameter("@PriorityDateSet", DBNull.Value));
                }
                else
                {
                    Command.Parameters.Add(new SqlParameter("@PriorityDateSet", oJobOfWork.PriorityDateSet));
                }

                Command.ExecuteNonQuery();
            }

            public void Update_Status(int JobOfWorkID, int Status)
            {
                _database.ClearParameter();
                _database.AddParameter("@JobOfWorkID", JobOfWorkID, true);
                _database.AddParameter("@Status ", Status, true);
                _database.ExecuteSP_NO_Return("JobOfWork_Update_Status");
            }

            public void Update_Priority(JobOfWork Jow)
            {
                _database.ClearParameter();
                _database.AddParameter("@JobOfWorkID", Jow.JobOfWorkID, true);
                _database.AddParameter("@Priority ", Jow.Priority, true);
                _database.AddParameter("@PriorityDateSet ", Jow.PriorityDateSet, true);
                _database.AddParameter("@QualificationID ", Jow.QualificationID, true);
                _database.ExecuteSP_NO_Return("JobOfWork_Update_Priority");
            }

            public DataView JobOfWork_Get_For_Job(int JobID)
            {
                _database.ClearParameter();
                _database.AddParameter("@JobID", JobID, true);
                var dt = _database.ExecuteSP_DataTable("JobOfWork_Get_For_Job");
                dt.TableName = Enums.TableNames.tblJobOfWork.ToString();
                return new DataView(dt);
            }

            public DataView JobOfWork_Get_For_Job(int JobID, SqlTransaction trans)
            {
                var Command = new SqlCommand();
                Command.CommandText = "JobOfWork_Get_For_Job";
                Command.CommandType = CommandType.StoredProcedure;
                Command.Transaction = trans;
                Command.Connection = trans.Connection;
                Command.Parameters.AddWithValue("@JobID", JobID);
                var Adapter = new SqlDataAdapter(Command);
                var returnDS = new DataSet();
                Adapter.Fill(returnDS);
                var dt = returnDS.Tables[0];
                dt.TableName = Enums.TableNames.tblJobOfWork.ToString();
                return new DataView(dt);
            }

            public DataTable JobOfWork_Get_ForEngineerVisitID(int engineerVisitId)
            {
                var dt = new DataTable();
                var command = new SqlCommand("JobOfWork_Get_ForEngineerVisitID", new SqlConnection(_database.ServerConnectionString));
                try
                {
                    command.CommandType = CommandType.StoredProcedure;
                    command.Parameters.AddWithValue("@EngineerVisitID", engineerVisitId);
                    dt = _database.ExecuteCommand_DataTable(command);
                    dt.TableName = "tblJobOfWork";
                    return dt;
                }
                catch (Exception ex)
                {
                    return dt;
                }
                finally
                {
                    command.Connection.Close();
                }
            }

            public JobOfWork JobOfWork_Get(int JobOfWorkID)
            {
                _database.ClearParameter();
                _database.AddParameter("@JobOfWorkID", JobOfWorkID, true);
                var dt = _database.ExecuteSP_DataTable("JobOfWork_Get");
                JobOfWork jobOfWork = null;
                if (dt.Rows.Count > 0)
                {
                    jobOfWork = new JobOfWork();
                    {
                        var withBlock = dt.Rows[0];
                        jobOfWork.IgnoreExceptionsOnSetMethods = true;
                        jobOfWork.Exists = true;
                        jobOfWork.SetJobOfWorkID = withBlock["JobOFWorkID"];
                        jobOfWork.SetJobID = withBlock["JobID"];
                        jobOfWork.SetPONumber = withBlock["PONumber"];
                        jobOfWork.SetDeleted = Conversions.ToBoolean(withBlock["Deleted"]);
                        jobOfWork.SetStatus = Helper.MakeIntegerValid(withBlock["Status"]);
                        jobOfWork.SetPriority = Helper.MakeIntegerValid(withBlock["Priority"]);
                        jobOfWork.PriorityDateSet = Helper.MakeDateTimeValid(withBlock["PriorityDateSet"]);
                        if (dt.Rows[0].Table.Columns.Contains("QualificationID"))
                            jobOfWork.SetQualificationID = Helper.MakeIntegerValid(withBlock["QualificationID"]);
                    }
                }

                return jobOfWork;
            }

            public JobOfWork JobOfWork_Get_As_Object(int jobOfWorkID)
            {
                _database.ClearParameter();
                _database.AddParameter("@JobOfWorkID", jobOfWorkID, true);
                var dt = _database.ExecuteSP_DataTable("JobOfWork_Get");
                JobOfWork jobOfWork = null;
                if (dt.Rows.Count > 0)
                {
                    jobOfWork = new JobOfWork();
                    {
                        var withBlock = dt.Rows[0];
                        jobOfWork.IgnoreExceptionsOnSetMethods = true;
                        jobOfWork.Exists = true;
                        jobOfWork.SetJobOfWorkID = withBlock["JobOFWorkID"];
                        jobOfWork.SetJobID = withBlock["JobID"];
                        jobOfWork.SetPONumber = withBlock["PONumber"];
                        jobOfWork.SetDeleted = Conversions.ToBoolean(withBlock["Deleted"]);
                        jobOfWork.SetStatus = Helper.MakeIntegerValid(withBlock["Status"]);
                        jobOfWork.SetPriority = Helper.MakeIntegerValid(withBlock["Priority"]);
                        jobOfWork.PriorityDateSet = Helper.MakeDateTimeValid(withBlock["PriorityDateSet"]);
                        if (dt.Rows[0].Table.Columns.Contains("QualificationID"))
                            jobOfWork.SetQualificationID = Helper.MakeIntegerValid(withBlock["QualificationID"]);
                        jobOfWork.JobItems = _database.JobItems.JobOfWork_Get_For_Job_Of_Work_As_Objects(jobOfWork.JobOfWorkID);
                        jobOfWork.EngineerVisits = _database.EngineerVisits.EngineerVisits_Get_For_Job_Of_Work_As_Objects_Light(jobOfWork.JobOfWorkID);
                    }
                }

                return jobOfWork;
            }

            public ArrayList JobOfWork_Get_For_Job_As_Objects(int JobID, SqlTransaction trans)
            {
                var jobOfWorks = new ArrayList();
                foreach (DataRow row in JobOfWork_Get_For_Job(JobID, trans).Table.Rows)
                {
                    var jobOfWork = new JobOfWork();
                    jobOfWork.IgnoreExceptionsOnSetMethods = true;
                    jobOfWork.Exists = true;
                    jobOfWork.SetJobOfWorkID = row["JobOFWorkID"];
                    jobOfWork.SetJobID = row["JobID"];
                    jobOfWork.SetPONumber = row["PONumber"];
                    jobOfWork.SetDeleted = Conversions.ToBoolean(row["Deleted"]);
                    jobOfWork.SetStatus = Helper.MakeIntegerValid(row["Status"]);
                    jobOfWork.SetPriority = Helper.MakeIntegerValid(row["Priority"]);
                    jobOfWork.PriorityDateSet = Helper.MakeDateTimeValid(row["PriorityDateSet"]);
                    if (row.Table.Columns.Contains("QualificationID"))
                        jobOfWork.SetQualificationID = Helper.MakeIntegerValid(row["QualificationID"]);
                    jobOfWork.JobItems = _database.JobItems.JobOfWork_Get_For_Job_Of_Work_As_Objects(jobOfWork.JobOfWorkID, trans);
                    jobOfWork.EngineerVisits = _database.EngineerVisits.EngineerVisits_Get_For_Job_Of_Work_As_Objects(jobOfWork.JobOfWorkID, trans);
                    jobOfWorks.Add(jobOfWork);
                }

                return jobOfWorks;
            }

            public ArrayList JobOfWork_Get_For_Job_As_Objects(int JobID)
            {
                var jobOfWorks = new ArrayList();
                foreach (DataRow row in JobOfWork_Get_For_Job(JobID).Table.Rows)
                {
                    var jobOfWork = new JobOfWork();
                    jobOfWork.IgnoreExceptionsOnSetMethods = true;
                    jobOfWork.Exists = true;
                    jobOfWork.SetJobOfWorkID = row["JobOFWorkID"];
                    jobOfWork.SetJobID = row["JobID"];
                    jobOfWork.SetPONumber = row["PONumber"];
                    jobOfWork.SetDeleted = Conversions.ToBoolean(row["Deleted"]);
                    jobOfWork.SetStatus = Helper.MakeIntegerValid(row["Status"]);
                    jobOfWork.SetPriority = Helper.MakeIntegerValid(row["Priority"]);
                    jobOfWork.PriorityDateSet = Helper.MakeDateTimeValid(row["PriorityDateSet"]);
                    if (row.Table.Columns.Contains("QualificationID"))
                        jobOfWork.SetQualificationID = Helper.MakeIntegerValid(row["QualificationID"]);
                    jobOfWork.JobItems = _database.JobItems.JobOfWork_Get_For_Job_Of_Work_As_Objects(jobOfWork.JobOfWorkID);
                    jobOfWork.EngineerVisits = _database.EngineerVisits.EngineerVisits_Get_For_Job_Of_Work_As_Objects_Light(jobOfWork.JobOfWorkID);
                    jobOfWorks.Add(jobOfWork);
                }

                return jobOfWorks;
            }

            public void Update_PONumberByJobOfWorkID(int jobOfWorkId, string poNumber)
            {
                _database.ClearParameter();
                _database.AddParameter("@JobOfWorkID ", jobOfWorkId, true);
                _database.AddParameter("@PONumber ", poNumber, true);
                _database.ExecuteSP_NO_Return("JobOfWork_Update_PONumberByJobOfWorkID");
            }

            public List<JobOfWork> Get_ByJobId(int jobId)
            {
                _database.ClearParameter();
                _database.AddParameter("@JobID", jobId, true);
                var dt = _database.ExecuteSP_DataTable("JobOfWork_Get_For_Job");
                if (dt is object && dt.Rows.Count > 0)
                {
                    var jobOfWorks = (from x in dt.AsEnumerable()
                                      select new JobOfWork()
                                      {
                                          Exists = true,
                                          SetJobOfWorkID = Helper.MakeIntegerValid(x["JobOfWorkId"]),
                                          SetJobID = Helper.MakeIntegerValid(x["JobId"]),
                                          SetPONumber = Helper.MakeStringValid(x["PONumber"]),
                                          SetPriority = Helper.MakeIntegerValid(x["Priority"]),
                                          SetStatus = Helper.MakeIntegerValid(x["Status"]),
                                          PriorityDateSet = Helper.MakeDateTimeValid(x["PriorityDateSet"]),
                                          SetQualificationID = Helper.MakeIntegerValid(x["QualificationID"])
                                      }).ToList();
                    return jobOfWorks;
                }
                else
                {
                    return null;
                }
            }

            /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        }
    }
}